### PR TITLE
UefiCpuPkg: Fixing page table handling for non-present pages. [202405]

### PR DESCRIPTION
**Pull request by apop5**
## Description

This change will add logic to populate the address and propagate bits for the non-present pages.

Without this change, the issue can be seen as when a smaller page is marked present from a big non-present page, the remainder of the non-present pages will return `EFI_UNSUPPORTED` when using get attributes function though protocols like CPU arch.

4ca95f223114

- [x] Impacts functionality?
- **Functionality** - Does the change ultimately impact how firmware functions?
- Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
- **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter validation improvement, ...
- [ ] Breaking change?
- **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
- Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
- **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
- Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested
This was tested on QEMU Q35 platform and booted to OS. N/A

## Integration Instructions
N/A

### CODE REVIEW

File: UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c [hunk 1]
```diff
@@ -390,6 +390,17 @@ PageTableLibMapInLevel (
       if (RETURN_ERROR (Status)) {
         return Status;
       }
+
+      OneOfPagingEntry.Pnle.Uint64 = 0;
+      // MU_CHANGE Start - Populate base address bits for non present pages
```
Comment (makubacki) [inline UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c:394]: The MU_CHANGE section should move up a line.

File: UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c [hunk 2]
```diff
@@ -390,6 +390,17 @@ PageTableLibMapInLevel (
       if (RETURN_ERROR (Status)) {
         return Status;
       }
+
+      OneOfPagingEntry.Pnle.Uint64 = 0;
+      // MU_CHANGE Start - Populate base address bits for non present pages
+      // TODO: Author unit test to verify the split pages meet expected attribute bits.
```
Comment (makubacki) [inline UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c]: Is there an issue or something else tracking this todo?

Comment (apop5) [reply UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c]: Not that I am aware of, I can open one.

File: UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c (hunk 1 above)
Comment (apop5) [reply UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c:394]: Updated in latest.

File: UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c (hunk 2 above)
Comment (apop5) [reply UefiCpuPkg/Library/CpuPageTableLib/CpuPageTableMap.c]: Removed TODO statement, and opened #1082 to track TODO statement.

**Review by os-d — APPROVED** [review]
Is this a 2311 cherry-pick? If so can you add which commit it was cherry-picked from?

**Comment by makubacki** [comment]
The old PR link could be removed from the commit subject. This commit will link to this PR which will explain it was a cherry-pick. Going directly to the original PR via that link might be confusing.

**Review by makubacki — APPROVED** [review]

**Pull request merged**